hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OMManagerClientVersionValidations.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.protocolPB;
+
+import org.apache.hadoop.ozone.ClientVersions;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos;
+
+/**
+ * OMManagerClientVersionValidations is a collection of validations to run for
+ * OM RPC calls and Client version compatibility.
+ */
+public final class OMManagerClientVersionValidations {
+  private OMManagerClientVersionValidations() {
+
+  }
+
+  public static void validateClientVersionPostProcess(
+      OzoneManagerProtocolProtos.Type lookupKey,
+      OzoneManagerProtocolProtos.OMRequest request,
+      OzoneManagerProtocolProtos.InfoBucketResponse infoBucketResponse) {
+    // ToDo: Add EC specific rejection of request based on key info.
+    if (request.hasVersion()
+        &&
+        ClientVersions.isClientCompatible(
+            ClientVersions.CLIENT_EC_CAPABLE,
+            request.getVersion())){
+      return;
+    }
+  }
+  public static void validateClientVersionPostProcess(
+      OzoneManagerProtocolProtos.Type lookupKey,
+      OzoneManagerProtocolProtos.OMRequest request,
+      OzoneManagerProtocolProtos.LookupKeyResponse lookupKeyResponse) {
+    // ToDo: Add FSO specific rejection of request based on bucket info.
```

Comment:
FSO client version validation is required during pre-process stage of request handling. 

**Expected behavior of pre-FSO client** : Only a new client can talk to FSO bucket.
The proposed idea is to to disallow any operations(read/list/delete/rename/update) on a FILE_SYSTEM_OPTIMIZED("FSO") bucket by a pre-FSO client(old version client). Need to add logic to throw exception when a pre-FSO client tries to access FSO bucket.

Can we add `validateClientVersionPreProcess()` function and invoke it in a single place(common) instead of repeating for each request type inside switch case, where it can do client version matching ?

Reply:
How will the pre process know that the bucket is a FSO bucket. If the `Pre` step accesses the bucket then in the normal path, we will be accessing the same bucket information twice and this might be wasteful or the code path could have to be changed to pass the bucket info around? Also, the client might not have permission for the bucket in which case the error should be around access denied and not a compatibility error. Checking for compatibility post allows for other errors and checks to still function and add comptability as a layer on top.